go-controller/pkg/ovn/zone_interconnect/interconnect.go

```diff
@@ -0,0 +1,542 @@
+	}
+
+	return nil
+}
+
+// AddRemoteZoneNode creates the interconnect resources in OVN NB DB for the remote zone node
+func (zic *ZoneInterconnectHandler) AddRemoteZoneNode(node *corev1.Node) error {
+	klog.Infof("Creating interconnect resources for remote zone node [%s] for the network [%s]", node.Name, zic.GetNetworkName())
+
+	nodeID := util.GetNodeID(node)
+	if nodeID == -1 {
+		// Don't consider this node as cluster-manager has not allocated node id yet.
+		return fmt.Errorf("failed to get node id for node - %s", node.Name)
+	}
+
+	// Get the chassis id.
+	chassisId, err := util.ParseNodeChassisIDAnnotation(node)
+	if err != nil {
+		return fmt.Errorf("failed to parse node chassis-id for node - %s, error: %v", node.Name, err)
+	}
+
+	if err := zic.createRemoteZoneNodeResources(node, nodeID, chassisId); err != nil {
+		klog.Errorf("Creating interconnect resources for remote zone node [%s] for the network [%s] failed : err - %v", node.Name, zic.GetNetworkName(), err)
+		return fmt.Errorf("creating interconnect resources for remote zone node [%s] for the network [%s] failed : err - %v", node.Name, zic.GetNetworkName(), err)
+	}
+
+	return nil
+}
+
+// DeleteNode deletes the local zone node or remote zone node resources
+func (zic *ZoneInterconnectHandler) DeleteNode(node *corev1.Node) error {
+	klog.Infof("Deleting interconnect resources for the node [%s] for the network [%s]", node.Name, zic.GetNetworkName())
+
+	// Cleanup the logical router port in the cluster router for the node
+	// if it exists.
+	if err := zic.cleanupNodeClusterRouterPort(node); err != nil {
+		return err
+	}
+
+	// Cleanup the logical switch port in the transit switch for the node
+	// if it exists.
+	return zic.cleanupNodeTransitSwitchPort(node)
+}
+
+// Cleanup deletes the transit switch for the network
+func (zic *ZoneInterconnectHandler) Cleanup() error {
+	klog.Infof("Deleting the transit switch [%s] for the network [%s]", zic.networkTransitSwitchName, zic.GetNetworkName())
+	return libovsdbops.DeleteLogicalSwitch(zic.nbClient, zic.networkTransitSwitchName)
+}
+
+func (zic *ZoneInterconnectHandler) createLocalZoneNodeResources(node *corev1.Node, nodeID int) error {
+	nodeTransitSwitchPortIPs, err := util.ParseNodeTransitSwitchPortAddrs(node)
+	if err != nil || len(nodeTransitSwitchPortIPs) == 0 {
+		return fmt.Errorf("failed to get the node transit switch port ips for node %s: %v", node.Name, err)
+	}
+
+	networkId, err := util.ParseNetworkIDAnnotation(node, zic.GetNetworkName())
+	if err != nil {
+		return fmt.Errorf("failed to get the network id for the network %s on node %s: %v", zic.GetNetworkName(), node.Name, err)
+	}
+
+	var transitRouterPortMac net.HardwareAddr
+	var transitRouterPortNetworks []string
+	foundIPv4 := false
+	for _, ip := range nodeTransitSwitchPortIPs {
+		if utilnet.IsIPv4CIDR(ip) {
+			foundIPv4 = true
+			transitRouterPortMac = util.IPAddrToHWAddr(ip.IP)
+		} else {
+			if !foundIPv4 {
```

Comment:
+1 it may be worth adding a comment here. Also, if we ever had more than one ipv4, the mac addr would be derived from the last one. Is that what we wanted?